### DISCUSSION TITLE
[v2]When using the Tauri Icon to generate icons, it is always larger than other icons in Mac

### DISCUSSION THREAD

**Discussion by nongyehong** [tauri-apps/tauri#10999] [Q&A] [answered]
![QQ_1726296742405](https://github.com/user-attachments/assets/c5e8d18b-b04c-41c8-b83e-3ecef05bf176)
Do I need to generate icons separately for Mac and package them? I've tried to fix this problem after modifying the icon so that the icon has a transparent border, but it will be much smaller in windows or the system tray, what is the solution to it?

**Comment by FabianLars** [comment]
Yeah, that's one of the cross-platform problems i'm not sure how to solve.

On macOS the icon indeed needs transparent padding in the icon file itself. Can't remember if it also requires manual shadows or not 🤔 
Then there's iOS that doesn't support transparency in the icon.

If you only support macOS then simply replacing icon.icns with the macos-specific icon should work. If you're also targetting iOS then this _may_ get tricky. I can't remember what it uses but if it also uses the icon.icns then replacing that with the iOS icon, then running `tauri ios init` and then changing it back to the macOS icon should work (afaik the ios project icons will only be replaced on init).
The tray icon should be derived from the png, not the icns file so unless i missed some other use of the icns file you should be fine then.

----

If you have any ideas on how to solve this, whether it's on a config level, in `tauri icon` or in the docs, please tell me 🙃

**Reply by FabianLars** [reply]
Actually, i think we don't use the .icns file on iOS at all. Checking it rn...

**Reply by nongyehong** [reply]
> Yeah, that's one of the cross-platform problems i'm not sure how to solve.
> 
> On macOS the icon indeed needs transparent padding in the icon file itself. Can't remember if it also requires manual shadows or not 🤔 Then there's iOS that doesn't support transparency in the icon.
> 
> If you only support macOS then simply replacing icon.icns with the macos-specific icon should work. If you're also targetting iOS then this _may_ get tricky. I can't remember what it uses but if it also uses the icon.icns then replacing that with the iOS icon, then running `tauri ios init` and then changing it back to the macOS icon should work (afaik the ios project icons will only be replaced on init). The tray icon should be derived from the png, not the icns file so unless i missed some other use of the icns file you should be fine then.
> 
> If you have any ideas on how to solve this, whether it's on a config level, in `tauri icon` or in the docs, please tell me 🙃

I think there should be an api to set all of the mac ICONS or set the Dock icon individually, because not only the Dock icon will be bigger, but also the app Center display will be bigger.🤔

**Comment by nongyehong** [comment, marked as answer]
You can refer to the previous discussion to lead to the solution: https://github.com/tauri-apps/tauri/discussions/11040, because the Mac icon needs to be configured separately, so we can set the icon for the Mac separately, but first make sure that your Mac icon follows the Mac icon standard: https://developer.apple.com/cn/design/human-interface-guidelines/app-icons#macOS

**Reply by luckbyte** [reply]
The original picture size should be 1024x1024, the actual picture should be 826x826, the excess space is blank。
<img width="943" alt="image" src="https://github.com/user-attachments/assets/060c173c-478f-4ba7-a148-3e208b8d6a86" />

**Reply by nongyehong** [reply]
🤔If you need to display it on mac, this is normal, because mac design specifications are like this.

**Reply by nongyehong** [reply]
![image](https://github.com/user-attachments/assets/b7b38ff7-ed47-4c3c-9dc1-851c2c80e2d4)

**Reply by luckbyte** [reply]
Yes, I wrote it because I wanted to help others with the same problem, the Apple website does not mention 826x826

**Reply by setoelkahfi** [reply]
So, is the solution is to create a macOS only config file and put the icon there?

**Reply by nongyehong** [reply]
> So, is the solution is to create a macOS only config file and put the icon there?

Yes, you need to create a separate folder with the mac icons configured separately, and then read the icons in your folder in tauri.macos.conf.json

**Reply by setoelkahfi** [reply]
I might not be following clearly. What's the problem with replacing the `icon.icns` with the square one?

**Reply by nongyehong** [reply]
> I might not be following clearly. What's the problem with replacing the `icon.icns` with the square one?

<img width="313" height="507" alt="image" src="https://github.com/user-attachments/assets/d152fb12-47d3-4fec-bbf6-f6e2e4fca973" />
<img width="793" height="703" alt="image" src="https://github.com/user-attachments/assets/81e9fa22-f750-4af1-9e1a-1b5de6a96102" />

I don't understand that my problem is related to the square icon you set. My problem is to solve the icon size problem.

**Reply by setoelkahfi** [reply]
Ah, ok. 

Now I see the problem with my solution of changing the `icon.icns` file directly with the square one:
```
2025-08-20 15:19:22.431 *** Error: ERROR: [ContentDelivery.Uploader] Validation failed (409) Missing required icon. The application bundle does not have an icon in ICNS format containing a 512pt x 512pt @2x image. For further assistance, see the Human Interface Guidelines at https://developer.apple.com/design/human-interface-guidelines/foundations/app-icons/. (ID: ece3e2b9-8d99-4062-ad76-18c88bde0d9e)
2025-08-20 15:20:06.052  INFO: [ContentDelivery.Uploader]
=============
UPLOAD FAILED with 1 error.
=============
2025-08-20 15:20:06.059 *** Error: Error uploading 'NotWallet.pkg'.
2025-08-20 15:20:06.059 *** Error: Validation failed Missing required icon. The application bundle does not have an icon in ICNS format containing a 512pt x 512pt @2x image. For further assistance, see the Human Interface Guidelines at https://developer.apple.com/design/human-interface-guidelines/foundations/app-icons/. (ID: ece3e2b9-8d99-4062-ad76-18c88bde0d9e) (409)
 {
    NSLocalizedDescription = "Validation failed";
    NSLocalizedFailureReason = "Missing required icon. The application bundle does not have an icon in ICNS format containing a 512pt x 512pt @2x image. For further assistance, see the Human Interface Guidelines at https://developer.apple.com/design/human-interface-guidelines/foundations/app-icons/. (ID: ece3e2b9-8d99-4062-ad76-18c88bde0d9e)";
    NSUnderlyingError = "Error Domain=IrisAPI Code=-19241 \"Validation failed\" UserInfo={status=409, detail=Missing required icon. The application bundle does not have an icon in ICNS format containing a 512pt x 512pt @2x image. For further assistance, see the Human Interface Guidelines at https://developer.apple.com/design/human-interface-guidelines/foundations/app-icons/., id=ece3e2b9-8d99-4062-ad76-18c88bde0d9e, code=STATE_ERROR.VALIDATION_ERROR, title=Validation failed, NSLocalizedFailureReason=Missing required icon. The application bundle does not have an icon in ICNS format containing a 512pt x 512pt @2x image. For further assistance, see the Human Interface Guidelines at https://developer.apple.com/design/human-interface-guidelines/foundations/app-icons/., NSLocalizedDescription=Validation failed}";
    "iris-code" = "STATE_ERROR.VALIDATION_ERROR";
}
```

**Reply by setoelkahfi** [reply]
So what are the sizes for your macOS icns icon? I put 1024x1024 png and simply renamed it to icon.icns, it fits perfectly on debug app. but got those error while uploading it to the appstore.

**Reply by FabianLars** [reply]
https://github.com/tauri-apps/tauri/blob/dev/crates/tauri-cli/src/helpers/icns.json these are the sizes `tauri icon` uses

**Reply by SeedyROM** [reply]
https://github.com/orgs/tauri-apps/discussions/10999#discussioncomment-15549464

This answer is what works best if you don't want to get super manual.

**Comment by LorenzoBloedow** [comment]
For people trying to do this with Icon Composer, make sure you export with this option and change the file extension to `icns`:

<img width="550" height="412" alt="Screenshot 2025-12-30 at 4 38 25 PM" src="https://github.com/user-attachments/assets/829406bf-32f3-47c9-8516-077a7abd74ac" />

**Comment by olimexsmart** [comment]
Still something is not clear to me: why this is not integrated into the `npm run tauri icon` script? 

Having worked with quite a few other app frameworks, IMHO this script that starts from a single `png` and just works, is just a dream. 

Making it perfect is adding a step where it resizes the icon fro macOS, adds rounded corners, copies the current tauri config, creates the macos folder and that's it

**Reply by FabianLars** [reply]
I tried that in https://github.com/tauri-apps/tauri/pull/11037 but couldn't make it not look super uncanny yet.

> IMHO this script that starts from a single png and just works, is just a dream.

And a very unrealistic dream because resizing a single png to these very different sizes typically looks horrible on at least one end of the spectrum. Of course easily fixable by just giving devs the option to provide more than one input icon but i still didn't get to that.

**Reply by olimexsmart** [reply]
First and foremost: thanks for your work. 

I think that most people would be happy to have a _decent_ icon on every OS. At the moment on macOS is not acceptable, it has to be fixed by hand, and that's a pity because the script is otherwise great. If an app is being developed for a project with tons of manpower and resources, the use of the `npm run tauri icon` is out of scope, they will prefer to personalize by hand for each OS anyway. Just a personal opinion.

**Comment by setoelkahfi** [comment]
I've published 10+ apps on the Mac AppStore with this semi-autonomous steps to generate macOS `icns` file. Assuming you already have the 1024x1024 app icon you use for running `cargo tauri icon` :

1. Convert to macOS layout here: https://www.candyicons.com/free-tools/app-icon-assets-generator
2. Download the files, unzip, and pick the `1024-mac.png`.
3. Then generate `icns` icon [here](https://img2icns.com/).
4. Download the file and replace the one in the `src-tauri/icons/icon.icns`.

**Reply by olimexsmart** [reply]
Thanks, just awesome. This should be the accepted answer.

**Reply by giray123** [reply]
I have just tried this and it does not work. The icon is still bigger without that selection paddings as well.

**Reply by timwehrle** [reply]
So for me it did work, but I need to find a way to make it more visually appealing besides the other MacOS icons. Especially the glass effect.

**Reply by SeedyROM** [reply]
This worked for me as well, thank you so much for helping me find this solution. GOAT'd! 🐐 🔥

**Reply by SohamPrabhudesai** [reply]
> So for me it did work, but I need to find a way to make it more visually appealing besides the other MacOS icons. Especially the glass effect.

Did you find a way to make it better?

**Reply by timwehrle** [reply]
@SohamPrabhudesai For the glass effect you probably need the icon composer of MacOS. That's where I ended
